Mathlib/NumberTheory/EulerProduct/Basic.lean

```diff
@@ -0,0 +1,170 @@
+/-
+Copyright (c) 2023 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import Mathlib.NumberTheory.SmoothNumbers
+import Mathlib.Analysis.SpecificLimits.Normed
+import Mathlib.Analysis.Normed.Field.InfiniteSum
+
+/-!
+# Euler Products
+
+The main results in this file are `EulerProduct.euler_product`, which says that
+if `f : ℕ → F` is norm-summable, where `F` is a complete normed field and `f` is
+multiplicative on coprime arguments with `f 0 = 0`, then
+`∏ p in primesBelow N, ∑' e : ℕ, f (p^e)` tends to `∑' n, f n` as `N` tends to infinity,
+and `EulerProduct.euler_product_multiplicative`, which states that
+`∏ p in primesBelow N, (1 - f p)⁻¹` tends to `∑' n, f n` as `N` tends to infinity,
+when `f` is completely multiplicative (implemented as `f : ℕ →*₀ F`).
+
+An intermediate step is `EulerProduct.summable_and_hasSum_smoothNumbers_prod_primesBelow_tsum`
+(and its variant `EulerProduct.summable_and_hasSum_smoothNumbers_prod_primesBelow_tsum_geometric`),
+which relates the finite product over primes to the sum of `f n` over `N`-smooth `n`.
+
+## Tags
+
+Euler product
+-/
+
+variable {F : Type*} [NormedField F] [CompleteSpace F]
```

Comment:
Do you actually use the fact that F is a field, for the general results on weakly (not completely) multiplicative functions? Would any commutative complete normed ring work? (I very much doubt the additional generality would be useful for anything, of course.)

Reply:
It seems that `norm_mul` and `norm_pow` need a division ring (and it has to be commutative for other reasons). This is maybe an artifact of the way a norm is defined. Possibly one can get by with just using upper bounds instead of equalities; I'll check.

Reply:
> It seems that norm_mul and norm_pow need a division ring

That's kind of annoying – it would be nice if there was a type class for "ring with a _multiplicative_ [semi?]-norm" – but this is not the place to fix it. I definitely don't think it's worth adding extra work and complication to the proofs for no obvious application.

Reply:
It was not that painful (I had to add a line using `Summable.of_nonneg_of_le` in two places), so I have reduced the requirements to the target being a complete normed commutative ring for the general statement. It then made sense to re-arrange the material so that everything related to the general statement comes  before everything related to the statement for completely multiplicative functions. Unfortunately, this makes the diff fairly large.